[쇼핑 주문 1, 2 단계] 크림 미션 제출합니다

ijh1298 opened the pull request:
안녕하세요, 두루! 소중한 시간 내어 코드 리뷰해주셔서 감사합니다.

부족한 부분이 정말 많았어요.. 리팩터링에 매진하겠습니다!!

[42ba4db](https://github.com/woowacourse/android-shopping-order/pull/105/commits/42ba4db25007ca5c3bf715e27d27e03b4b407ca9) 커밋부터 리뷰 부탁드립니다!

# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [ ] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [ ] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

- 직렬화 라이브러리로는 Gson을 사용했습니다. 빌드 타임에 검증하고 속도가 더 빠른 Kotlinx.Serialization을 적용하고 싶었으나, 미션 구현하는 데 많은 시간이 사용하느라 이전 미션 코드에서 사용하던 라이브러리를 계속 쓰게 되었습니다. 

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

https://github-production-user-asset-6210df.s3.amazonaws.com/192606356/448760773-d9c7972f-cfd1-47fe-9080-a4d59c1ca367.mp4

![]()

ijh1298 committed:
feat: set up the project

wondroid-world committed:
docs: 기능 요구 사항 작성

wondroid-world committed:
build: shimmer 의존성 추가

wondroid-world committed:
feat: 상품 목록 shimmer layout 구현

wondroid-world committed:
feat: recyclerView Item에 상품 목록 shimmer 추가

ijh1298 committed:
feat: recyclerView LoadingStateProduct의 ViewHolder, Adapter 적용

ijh1298 committed:
feat: 카탈로그 화면에 Shimmer UI 추가

ijh1298 committed:
feat: 장바구니 화면 로딩 UI 추가

- 장바구니 화면에 Shimmer Effect를 활용한 로딩 UI를 추가했습니다.
- 로딩 중에는 ShimmerFrameLayout이 화면에 표시되고, 로딩이 완료되면 사라집니다.
- CatalogViewModel과 CartViewModel에 2초의 임의 딜레이를 추가하여 로딩 효과를 확인할 수 있도록 했습니다.

wondroid-world committed:
feat: 사용자 인증 정보 저장

wondroid-world committed:
build: retrofit 의존성 추가

wondroid-world committed:
feat: 장바구니 아이템 DTO 추가

wondroid-world committed:
feat: Product service, Product retrofit 구현

ijh1298 committed:
feat: RemoteCatalogProductRepositoryImpl 구현 및 적용

- RemoteCatalogProductRepositoryImpl 추가
- CatalogProductRepository에 getProductsByPage 추가
- CatalogViewModel에 RemoteCatalogProductRepositoryImpl 의존성 추가
- LocalDummyCatalogProductRepositoryImpl 주석 처리
- network_security_config에 서버 도메인 추가
- ShoppingApplication에서 ID 변경

ijh1298 committed:
feat: Retrofit 서비스 싱글톤으로 변경 및 로깅 인터셉터 추가

ijh1298 committed:
refactor: 잘못 사용한 DTO import 변경

wondroid-world committed:
feat: 서버에서 상품 목록 불러오는 기능 구현

wondroid-world committed:
feat: 서버에서 상세 페이지 조회 구현

wondroid-world committed:
feat: detailActivty에서 최근 본 상품 조회 구현

wondroid-world committed:
feat: CartItem service, retrofit 구현

ijh1298 committed:
refactor: CartProductRepository를 Remote로 변경

wondroid-world committed:
feat: 장바구니에 상품 추가 후, cartItemId 콜백 받도록 구현

ijh1298 committed:
feat: 장바구니에 담긴 상품 수량 표시 및 상품 상세 화면으로 이동 시 상품 정보 전달 기능 개선

- CatalogViewModel에서 장바구니에 담긴 상품의 cartItemId를 ProductUiModel에 추가
- DetailViewModel에서 상품 정보를 productId 대신 ProductUiModel 객체로 받도록 수정
- DetailViewModel에서 장바구니에 상품 추가 시 이미 담겨있는 상품이면 수량 업데이트, 아니면 새로 추가
- DetailActivity에서 DetailViewModelFactory에 productId 대신 ProductUiModel 객체 전달
- CatalogActivity에서 상품 클릭 시 DetailActivity로 productId 대신 ProductUiModel 객체 전달
- 최근 본 상품 클릭 시 DetailActivity로 productId 대신 ProductUiModel 객체 전달

wondroid-world committed:
feat: 제품 상세 페이지에서 장바구니에 없는 상품을 추가하면, 서버에 연동되는 로직 구현

wondroid-world committed:
feat: 상품 목록 페이지 수량 증가 감소 버튼 로직 구현

ijh1298 committed:
fix: 상품 수량 1개일 때 장바구니에서 상품 삭제되도록 수정

ijh1298 committed:
feat: 상품 수량이 0인 경우 카트에 상품 추가, 0이 아닌 경우 수량 변경 기능 추가

ijh1298 committed:
refactor: 불필요한 레포지토리 및 객체 코드 제거

- CartProductRepositoryImpl 삭제
- HttpCatalogProductRepositoryImpl 삭제
- LocalDummyCatalogProductRepositoryImpl 삭제
- CartProductRepository의 getAllProductsSize 메서드 제거
- CartProductRepository의 getProductQuantity 메서드 제거
- RemoteCartProductRepositoryImpl의 불필요한 Log.d 및 println 제거

ijh1298 committed:
refactor: 불필요한 레포지토리 및 객체 코드 제거

- CartProductRepositoryImpl 삭제
- HttpCatalogProductRepositoryImpl 삭제
- LocalDummyCatalogProductRepositoryImpl 삭제
- CartProductRepository의 getAllProductsSize 메서드 제거
- CartProductRepository의 getProductQuantity 메서드 제거
- RemoteCartProductRepositoryImpl의 불필요한 Log.d 및 println 제거

ijh1298 committed:
Merge branch 'step1' of https://github.com/ijh1298/android-shopping-order into step1

wondroid-world committed:
docs: 2단계 기능 요구 사항 작성

wondroid-world committed:
feat: 주문 화면 레이아웃 구현

wondroid-world committed:
feat: 장바구니 상품 추천 fragment  layout 구현

wondroid-world committed:
feat: cart selection fragment 추가

ijh1298 committed:
fix: CartSelectionFragment 뷰 바인딩 오류 수정 및 레이아웃 조정

ijh1298 committed:
feat: 장바구니 상품 개별 선택 기능 추가

wondroid-world committed:
feat: cartActity에 cartViewModel 추가

ijh1298 committed:
feat: 장바구니 전체 금액, 총 개수 표시 기능 추가

ijh1298 committed:
fix: 장바구니 상품 선택 기능 및 총 금액 계산 기능 수정

ijh1298 committed:
feat: 장바구니에 담긴 상품이 없을 때 추천 상품 화면 표시 기능 추가

wondroid-world committed:
feat: 추천상품 받아올 수 있는 로직 구현

ijh1298 committed:
feat: 장바구니 추천 상품 UI 적용

ijh1298 committed:
refactor: CartActivity의 hasHandledTotalCount 불필요한 조건 제거

wondroid-world committed:
fix: 카트에 상품 목록이 없을 때, 상품 추천이 뜨도록 추천

Gyuil-Hwnag commented on README.md:
```diff
@@ -1,19 +1,18 @@
-- [x] 마지막으로 본 상품 페이지에서 뒤로 가기를 하면 상품 목록으로 이동한다.
-- [x] 앱이 재시작돼도 최근 본 상품 목록과 장바구니 데이터는 유지돼야 한다.
\ No newline at end of file
+- 스켈레톤 UI 노출
+  - [x] 상품 목록
+  - [x] 장바구니
```
Readme 작성 좋습니다👍
단계별이 아니라, 전체 내용을 수정하셨나 보군요~

Gyuil-Hwnag commented on app/build.gradle.kts:
```diff
@@ -62,10 +62,17 @@ dependencies {
     implementation(libs.okhttp)
     implementation(libs.gson)
     implementation(libs.mockwebserver)
+    implementation(libs.shimmer)
+    implementation(libs.retrofit)
+    implementation(libs.converter.gson)
+    implementation(libs.logging.interceptor)
+    implementation(libs.androidx.fragment.ktx)
```
아마 프로젝트를 설정할 때부터 version catalogs를 사용하도록 되어 있었을 것 같은데, 이전에는 gradle을 kotlin 이전에 groovy로 적던 시절도 있었어요~
오래된 프로젝트일 수록 마이그레이션이 이뤄지지 않은 경우 groovy, 또는 version catalogs 가 적용이 안된 프로젝트일 때도 있답니다!
이전에는 어떻게 gradle 을 정의했었는지도 추후에 시간이 남을 때 살펴본다면 많은 도움이 될 것 같다는 생각이 듭니다💪

지금 당장 하기에는 너무 과한 내용일 것 같아서 추후에 한번 시간이 남을 때 살펴보시면 좋을 것 같아요

- groovy -> groovy to kotlin -> buildSrc등 version catalogs 이전 gradle 관리법 -> version catalogs -> custom plugins

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/ShoppingApplication.kt:
```diff
@@ -9,10 +10,23 @@ class ShoppingApplication : Application() {
 
+    val sharedPreferences: SharedPreferences by lazy {
+        getSharedPreferences(USER_INFO, MODE_PRIVATE)
+    }
```
SharedPreference 를 사용해주셨군요~👍
한가지 고민이 드는 것은 Application에 변수가 있다면, 만약 모듈을 분리하는등의 이유로 Application에 대해서 알 수 없는 상태라면 어떻게 해볼 수 있을까요?!

ijh1298 replied:
이전 미션에서도 동일하게 Application에 선언하고 쓰고 있었는데, 사실 큰 문제를 느끼지 못했습니다.🙃
그래서 그때 리뷰어분이 준 질문에 대해서도 `Database를 Application에서 초기화하는 이유가 있냐?`는 질문에
생명주기와 관련해서 근거를 들어 대답했던 거 같아요. (초기화 시점은 그때가 맞다고 생각합니다!)

(두루의 피드백을 받고 느낀건데) 어쩌면 그때의 질문 의도가, Application에 종속된 변수로 선언됨으로써
> 만약 모듈을 분리하는등의 이유로 Application에 대해서 알 수 없는 상태라면 어떻게 해볼 수 있을까요?!
위와 같은 문제 상황이 발생할 수 있지 않을까하는 이야기였을 수도 있겠네요..!

결론적으로, `SharedPreference`, `Database`같은 변수를 외부에서 싱글톤처럼 접근할 수 있도록
`DependencyProvider`라는 object를 선언하여 관리하게 되었습니다.

그리고 ID, 비밀번호를 조금 더 안전하게 저장하는 BuildConfig 방식을 알게 되면서 `SharedPreference`는 제거했습니다.😅

[`feat: RepositoryProvider, DependencyProvider 추가 및 ViewModelFactory 적용`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/333c4bd1a6999ddc65509200db3ecb32ece0a93f)

Gyuil-Hwnag replied:
> SharedPreference, Database같은 변수를 외부에서 싱글톤처럼 접근할 수 있도록
DependencyProvider라는 object를 선언하여 관리하게 되었습니다.

아주 잘 의도를 파악하셨군요👍

> BuildConfig 방식

더 나아가 BuildConfig 방식의 적용도 정말 좋습니다~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/ShoppingApplication.kt:
```diff
@@ -9,10 +10,23 @@ class ShoppingApplication : Application() {
         database = ShoppingDatabase.getInstance(applicationContext)
+
+        sharedPreferences.edit().putString(ID, "ijh1298")
+        sharedPreferences.edit().putString(PASSWORD, "password")
```
직접 하드코딩으로 넣어주셨군요!
이렇게 직접 하드코딩을 하는 방법외에 어떤 방법들이 있을까요??🤔
ApiKey등으로 발급 받은 것을 예시로 들어보면, 이를 어떻게 관리할 수 있을까요

ijh1298 replied:
아이디, 비밀번호, API Key 등 민감한 정보를 하드코딩으로 하지 않고,

local.properties에 값을 선언해둔 후 BuildConfig를 통해 ID, 비밀번호를 참조하도록 수정했습니다!

[`refactor: ID, PW를 BuildConfig에서 가져오도록 수정 및 SharedPreference 제거`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/7f514b5184977c990db65a5d129d4e365128f627)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartActivity.kt:
```diff
@@ -3,83 +3,72 @@ package woowacourse.shopping.cart
+        )[CartViewModel::class.java]
+    }
+    private var hasHandledTotalCount = false
 
```
이를 Activity의 flag 값을 활용하도록 하셨군요!
이를 ViewModel 에서 관리를 하려면 어떠한 방법이 있을까요~?
살짝의 힌트를 드려보자면 MediatorLiveData 등도 있을 것 같다는 생각이 듭니다

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartActivity.kt:
```diff
@@ -3,83 +3,72 @@ package woowacourse.shopping.cart
+                    }
+                }
+            }
+        }
```
해당 내용을 어떻게 하면 좀 더 뎁스를 줄여볼 수 있을까요~?🤔
when + reture 등을 활용하는 방법등도 있을 것 같다는 생각이 들었어요

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartAdapter.kt:
```diff
@@ -12,6 +12,7 @@ class CartAdapter(
     private val onDeleteProductClick: DeleteProductClickListener,
     private val onPaginationButtonClick: PaginationButtonClickListener,
     private val quantityControlListener: QuantityControlListener,
+    private val onCheckClick: CheckClickListener,
```
억지스럽긴 하지만, 만약 클릭이 10개가 있다면, 모두 Adapter에 추가를 하실건가요~
크림의 의견을 남겨주세요!💪

ijh1298 replied:
저 역시 리팩터링이 필요하다고 인지하고 공감하는 문제였습니다!🤗

`CartAdapter`에는 적용이 안되어 있는데, `QuantityControlListener` 같은 경우 증가, 감소 하는 리스너를 공통적인 분모가 있다고 생각하여 `하나의 interface에 여러 추상 함수를 선언`해 놓았습니다. 두루가 말씀하신 질문에 대한 제 답은, **하나의 interface에 여러 함수를 선언**해두면 해결할 수 있다고 생각해요!

이번에 리팩터링을 진행하면서 Pagination 로직을 제거해야겠다고 결정하여 `onPaginationButtonClick`을 제거했습니다. `QuantityControlListener`는 다른 곳에서도 사용하기 때문에 굳이 한번 더 묶지 않았습니다. `onDelete`, `onCheck` 리스너는 CartAdapter에만 사용될 거라 생각해서, `CartActionListener` 라는 interface를 선언해 한 곳에 두었습니다!

[`refactor: CartActionListener 인터페이스로 삭제 및 체크 이벤트 통합`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/4468fde3a6fd6a24e0480a708fc50f7101cc79d0)

ijh1298 replied:
깜빡하고 장바구니 화면에 Pagination 로직을 제거한 이유를 남기지 않았었네요..!

장바구니 화면에 들어갔을 때, 모든 상품이 선택되어 있는게 자연스럽다고 생각했습니다. 상품들을 구매하기 위해 장바구니에 담는다고 생각이 들었어요. 따라서 기본적으로 전체 선택 체크가 걸려있고, 총합 가격이 나오는게 자연스러운 흐름이라 생각했어요.

그러면 결국 (pagination 없이 총합 가격과 총 선택된 개수를 주는 API가 있지 않는 한) 모든 상품을 다 로드하고,
계산이 필요하다고 판단했어요. 이미 다 로드한 상태라면 `Pagination의 의의가 사라지지 않을까?`하며 없애게 되었어요!

두루는 어떻게 생각하시는지 궁금합니다😀!!

Gyuil-Hwnag replied:
@ijh1298 
엇! 이 부분이 코멘트가 안남겨졌었군요~🫣
오호~ 그렇게 생각을 할 수 있을 것 같군요!
그치만, 개인적으로는 아래와 같은 생각도 들었어요
만약 무한 스크롤이라면, 어떻게 할 것 인가?!
(매번 새롭게 불러오게 된다면, 스크롤이 초기화 되는등 여러가지 문제가 발생할 수도 있을 것 같군요🥲)

개인적으로는 이럴 경우 아래와 같은 방법을 채택할 수도 있겠다는 생각이 듭니다!
결과에 대해서 성공이라면, 반영됐다는 것으로 간주하여 클라에서 copy등으로 객체를 변환한다등을 채택해볼 수 있을 것 같았어요~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartRecommendationFragment.kt:
```diff
@@ -0,0 +1,74 @@
+        ViewModelProvider(
+            requireActivity(),
+            CartViewModelFactory(requireActivity().application as ShoppingApplication),
+        )[CartViewModel::class.java]
```
위에 Application에 남겨두었던 코멘트를 고민하다보면, 매번 ViewModelFactory 에서 Application 으로 캐스팅 하는 내용은 보완이 될 수 있겠군요~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartRecommendationFragment.kt:
```diff
@@ -0,0 +1,74 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+    }
```
`onCreate`를 오버라이딩 해두신 이유가 있을까요~?

Gyuil-Hwnag replied:
추가로 Fragment 의 생명주기는 Activity 와는 다른데, Fragment 의 생명주기에 대한 크림의 의견을 남겨주세요!!💪

ijh1298 replied:
Fragment 파일을 생성하며 기본적으로 정의된 `onCreate` 메서드를 수정하지 않아서 남아있었어요.. 제거했습니다!

Fragment 생명주기는 호스팅된 Activity 늦게 생성되고, 더 일찍 제거 된다고 알고 있어요.

Activity의 `onCreate` - `onStart` - `onResume` - `onPause` - `onStop` - `onDestroy` 생명주기 중, 
Fragment는 `onCreateView`, `onDestroyView` 등 생명주기 메서드가 더 존재하는 것으로 알고 있습니다. 

`onCreateView`에서 inflater를 사용할 수 있게 됨으로 binding 객체 inflate를 이곳에서 진행했습니다!

작성하다보니 두루의 의도를 알게 된 거 같은데요(?), `onDestroyView`에서 Fragment에서 binding 해제를 하지 않으면 메모리 누수가 될 수 있다는 점, lifeCycleOwner을 Fragment의 전체 생명주기인 `this`보다 View의 생명주기인 `viewLifeCycleOwner`로 설정해두는 게 좋다는 점을 상기하게 되었어요!

[`refactor: Fragment onDestroyView에서 binding 해제`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/76c1e2943c8cb12bcf29a3bc7530d35278ace219)

Gyuil-Hwnag replied:
오호~ 사실 처음의 의도는 Fragment 의 생명주기를 알고 있는가? 에서 시작을 하였지만, 너무 잘 수정을 해주셨어요👍
제가 2차, 3차로 남길 내용들에 대해서도 파악해서 적용을 해주셨군요 정말 좋습니다💯

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartActivity.kt:
```diff
@@ -3,83 +3,72 @@ package woowacourse.shopping.cart
+                    supportFragmentManager.commit {
+                        setReorderingAllowed(true)
+                        replace(R.id.fragment_container_cart_selection, CartSelectionFragment())
+                    }
```
해당 부분은 따로 함수로 관리를 한다면, 중복적인 구현에 대해서 줄여볼 수 있겠군요~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartRecommendationFragment.kt:
```diff
@@ -0,0 +1,74 @@
+            ProductAdapter(
+                products = emptyList(),
+                productActionListener =
+                    object : ProductActionListener {
+                        override fun onProductClick(product: ProductUiModel) {
+                            val intent = DetailActivity.newIntent(requireContext(), product)
+                            startActivity(intent)
+                        }
```
ProductActionListener에 대해서 재 사용을 하려다 보니, 이와 같이 불필요하게 오버라이딩을 지정해야 하는 부분들도 생기게 됬나 보군요🥲

ijh1298 replied:
이를 해결할 제일 좋은 방법은 **인터페이스 분리 원칙을 잘 따른 설계**라는 생각이 듭니다..
사용하지 않는 함수를 재정의해야 하는 인터페이스를 애초 만들지 않는게 제일 좋다고 생각이 들어요!

다만 코드 변경 사항이 많이 생길 거 같아 다른 방식의 해결책을 찾아보았는데, 
`interface` 내 기본 구현을 Unit으로 설정해두니 보기에는 깔끔해지게 되어 적용해보았습니다!

```kotlin
interface ProductActionListener {
    fun onProductClick(product: ProductUiModel)

    fun onLoadButtonClick() = Unit // 기본 구현 Unit 설정
}
```

[`refactor: ProductActionListener의 onLoadButtonClick 기본 구현 추가`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/8b98e71715204cfa4d7f2c069c3679e783ab2c94)

Gyuil-Hwnag replied:
아주 좋습니다~
> 사용하지 않는 함수를 재정의해야 하는 인터페이스를 애초 만들지 않는게 제일 좋다고 생각이 들어요!

개인적으로도 이 의견에 동의를 합니다!
interface에 대해서 너무 큰 책임을 싫어주다 보면, 점점 관리 측면에서는 비효율적으로 변하는 것 같아서, interface로 만들 때에 크림의 기준을 잘 세워서 분리를 하는 것이 중요하겠어요

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartSelectionFragment.kt:
```diff
@@ -0,0 +1,98 @@
+                onDeleteProductClick =
+                    DeleteProductClickListener { product ->
+                        viewModel.deleteCartProduct(CartItem.ProductItem(product))
+                    },
```
처음에는 `DeleteProductClickListener` 으로 지정을 하는 부분은 없어도 된다에서 시작을 하게 되었는데, `CartItem.ProductItem(product)`로 객체 인스턴스를 생성하게 되는 구현이 나오게 된 이유가 궁금합니다!

ijh1298 replied:
ViewModel 내에 전체 장바구니 아이템을 관리하는 변수의 자료형이 `LiveData<List<CartItem.ProductItem>>`였던 듯해요. 초기 구현에서 같은 자료형을 전달해 삭제하려고 했던 시도에서 나온 결과 같습니다🙃

새로운 객체 인스턴스를 생성해서 전달하는 것도 이상하고, `delete` 같은 로직은 id 값으로도 수행할 수 있는데(최소 지식을 전달하는 게 바람직한데) 이해하기 힘든 코드를 작성했었네요. 수정했습니다!

[`refactor: CartItem 삭제 시 ProductUiModel 대신 cartItemId 사용`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/679bb71b32397a6928db29460e4a70f8f48fbbc8)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartViewModel.kt:
```diff
@@ -29,15 +31,62 @@ class CartViewModel(
     private val _updatedItem = MutableLiveData<ProductUiModel>()
     val updatedItem: LiveData<ProductUiModel> = _updatedItem
 
+    private val _loadingState: MutableLiveData<LoadingState> =
+        MutableLiveData(LoadingState.loading())
+    val loadingState: LiveData<LoadingState> get() = _loadingState
```
이 부분을 보다가 LoadingState 라는 것이 존재를 해야 하는 건가? 라는 생각이 들었습니다!
LoadingState 내의 생성자들에 대해서는 따로 사용을 안하고 있는데, 그러면 차라리 `_isLoading: MutableLiveData<Boolean>`으로 처리 해도 되지 않을까? 라는 생각이 들었어요

ijh1298 replied:
`LoadingState` 객체가 생긴 이유는 (사실 잘 모르고 사용한 것도 크지만) Shimmer를 사용하는 다른 사람의 블로그에서 loading 상태를 따로 관리해주는 것이 좋은 아이디어라고 생각이 들어서였어요!

두루 말씀대로 최초에는 `LoadingState` 객체의 활용성이 별로 없었어요.😓

리팩터링을 하다보니, 상품 목록을 최초에 불러올 때는 `shimmer`를 통해 `skeleton`을 보여주는 게 사용자 경험적으로 좋았는데, 상품 목록을 개별로 업데이트(eg. 수량 증감)할 때도 `skeleton`이 보이는 불편한 현상이 있었어요. 그럴 때(`최초 로딩은 아니고, 아이템 변화로 인한 로딩`)는 `shimmer`를 작동시키지 않는 상태가 필요하다는 생각이 들었고, 결과적으로 `refreshing`이라는 상태를 추가해서 존재감을 확대시켰습니다!

[`refactor: Cart 초기 로딩이 아닐 시 Shimmer 적용 구분`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/2af8680ad71105565fe46c47b4dd33634ed5e118)

Gyuil-Hwnag replied:
개인적으로 아래와 같은 변수가 사용이 유의미한가? 라는 생각이 들었어요!
1차적으로는 변수들이 꼭 필요한가? 필요가 없다면 enum, data object 등이 있었을 것 같고 이를 통해서 오히려 더 간결하고 명확하게 나타낼 수 있지 않았을까? 라는 생각이 들었습니다
~~~kotlin
    val isLoading: Boolean,
    val shimmerVisibility: Int,
    val recyclerViewVisibility: Int,
~~~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartViewModel.kt:
```diff
@@ -75,16 +124,41 @@ class CartViewModel(
+            if (product.quantity != 1) {
+                cartProductRepository.updateProduct(product, product.quantity - 1) { result ->
+                    if (result == true) {
+                        _updatedItem.postValue(product.copy(quantity = product.quantity - 1))
```
사소한 내용이긴 한데, `if (result) vs if (result == true)`두개중 `result == true` 형태로 하신 이유가 궁금합니다~
처음에 해당 부분을 봤을 때, 이렇게 정의되어 있는 것을 보고 result 가 nullable 한가? 라는 생각에서 시작을 했어요

ijh1298 replied:
callback으로 돌아오는 자료형이 `Boolean` 값임을 알기 힘들거 같아서 명시적으로 작성했습니다! 
is로 시작하는 네이밍도 아니다 보니 명확하게 해주고 싶은 느낌이 있었습니다.

저 역시 저번 미션부터 LiveData의 getValue를 통한 nullable 값을 계속해서 쓰다보니,
`== true` 검증이 null을 걸러내기 위할 때도 많이 쓰게 됨을 인식했어요!

Gyuil-Hwnag replied:
오호 좋습니다~ `nullable 값을 계속해서 쓰다보니 + 명시적으로 작성했습니다!` 라는 의견 아주 좋은 것 같습니다ㅎㅎ

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/dto/product/Pageable.kt:
```diff
@@ -0,0 +1,19 @@
+import com.google.gson.annotations.SerializedName
+
+data class Pageable(
+    @SerializedName("offset")
+    val offset: Long,
```
@SerializedName 키워드를 붙이면 어떤 동작을 하게 되는 걸까요~?🤔

ijh1298 replied:
JSON의 key-value 값을 제가 생성한 class에 맞출 때 쓰는 걸로 알고 있습니다!

```kotlin
data class Person(
    @SerializedName("name")
    val userName: String?,
    @SerializedName("age")
    val userAge: Int?,
)
```

위와 같이 gson 어노테이션을 붙여 정의한 객체가 있으면,  
gson이 역직렬화 할 때 `name`이라는 SerializedName으로 key 값에 해당하는 value를 `userName`에게 할당해줄 수 있습니다!

지금은 data class의 프로퍼티와 SerializedName 어노테이션들이 전부 동일한데, 명시적으로 작성하기 위함이었어요!

Gyuil-Hwnag replied:
`명시적으로 작성하기 위함` 이었다면 좋습니다~
해당 어노테이션에 대한 이해를 얼마나 하고 있는지 궁금해서 남겨두었던 코멘트 였습니다!
추후에 시간이 난다면, 해당 어노테이션으로 인해서 어떻게 값을 매핑하게 되는지도 알아본다면 도움이 될 것 같다는 생각이 들었습니다

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/repository/RemoteCartProductRepositoryImpl.kt:
```diff
@@ -0,0 +1,241 @@
+                    ),
+            ).enqueue(
+                object : Callback<Void> {
+                    override fun onResponse(
+                        call: Call<Void>,
+                        response: Response<Void>,
+                    ) {
+                        if (response.isSuccessful) {
+                            val locationHeader = response.headers()["location"]
+                            val id = locationHeader?.substringAfterLast("/")?.toIntOrNull()
+                            callback(cartProduct.copy(cartItemId = id))
+                        }
+                    }
+
+                    override fun onFailure(
+                        call: Call<Void>,
+                        t: Throwable,
+                    ) {
+                        println("error : $t")
+                    }
```
해당 내용에서는 아래의 두가지 내용에 대한 궁금증이 생겼습니다~
1. response.isSuccessful = false 일 경우에는 어떤 경우 일까요?
2. onFailure 는 언제 타게 되는 걸까요?
3. onFailure 로 실패를 하는 경우에 사용자는 해당 화면을 계속 기다리는 것일까요~?
ㄴ 이는 실제 서비스라고 생각을 해본다면, 갑자기 화면이 아무것도 안그려지는데 사용자는 계속해서 기다려야 할까요?! 크림이 실제 앱을 사용할 때 이런 상황이 발생한다면 어떤 생각이 드실까요

ijh1298 replied:
해당 코드들의 내부 주석을 통해 다음과 같은 사실을 알게 되었어요.💡

>1. response.isSuccessful = false 일 경우에는 어떤 경우 일까요?

isSuccessful이 true일 때가 `response.code()`가 [200..300) 범위에 속하는 경우입니다.

>2. onFailure 는 언제 타게 되는 걸까요?

서버와 통신하는 동안 네트워크 예외가 발생하거나 요청을 생성하거나 
응답을 처리하는 동안 예기치 않은 예외가 발생했을 때 호출됩니다.

>3. onFailure 로 실패를 하는 경우에 사용자는 해당 화면을 계속 기다리는 것일까요~?

네트워크 연결에 실패했다는 문구를 띄우거나, 특정 횟수를 재시도 해보는 등 처리가 필요할 거 같아요.🥲

리뷰 요청이 너무 미루어지는 듯하여 이 부분을 비롯해 부족한 부분은 다음 리팩터링 때 진행해보겠습니다!!

Gyuil-Hwnag replied:
아주 좋습니다👍
해당 call의 상태가 어떤 상황에 발생하는지에 대해서 알아둔다면, 추후 개발을 할때에 이를 유의하면서 좀 더 유용하게 사용을 해볼 수 있을 것 같아서 남겨둔 내용이었습니다~

크림이 말해주신대로 부족한 부분은 리팩토링을 하면서 적용을 해봐도 좋을 것 같군요💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/repository/RemoteCatalogProductRepositoryImpl.kt:
```diff
@@ -0,0 +1,186 @@
+
+    override fun getCartProductsByUids(
+        uids: List<Int>,
+        callback: (List<ProductUiModel>) -> Unit,
```
아하 uid가 최근본 아이템 id 였군요😅
순간 보고 나서 userId 인가? 라는 생각을 했었군요

ijh1298 replied:
혼란을 줄 수 있는 uid라는 네이밍을 `productId`로 수정했어요!

[`refactor: uid, id를 productId로 이름 변경`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/f5bb0e3b97776be9c5bc8dc2e8d99afad5ba7368)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/service/CartItemService.kt:
```diff
@@ -0,0 +1,48 @@
+    @PATCH("/cart-items/{id}")
+    fun patchCartItemQuantity(
+        @Header("accept") accept: String = "*/*",
+        @Path("id") cartItemId: Int,
```
`@Header("accept") accept: String = "*/*"` 에 대해서 매번 API 가 추가될 때마다 이렇게 붙이지 않는 방법은 어떤 것이 있을까요~
(이 부분은 적용을 하는데 시간이 오래걸릴 것 같다면, 말씀주세요!)
힌트 : interceptor

ijh1298 replied:
Header 설정 부분도 실은 강의 자료의 것을 가져다 쓰느라 정확한 역할을 몰랐는데, 두루 덕분에 찾아보게 되었어요.💡

Header에서 `accept`을 통해 클라이언트가 서버에게 어떤 형태의 데이터를 응답으로 원하는지 명시하는군요! 
`*/*`은 JSON, XML, HTML 등 아무 형식이나 달라는 요청이었네요.

막연히 쓰고 있던 interceptor도 알아보니, `request`나 `response`를 중간에서 가로채고 처리하는 역할을 한다고 합니다.
지금처럼 `공통 헤더 추가`, `디버깅을 위한 로깅`, `토큰 갱신` 등을 위해 사용함을 알게 되었습니다.

매 요청때마다 Header에 붙이는 대신 Interceptor를 이용해 Header에 "accpet", "*/*"를 붙이도록 수정했습니다!

[`refactor: Retrofit Interceptor에서 Accept 헤더 추가 및 서비스 인터페이스에서 헤더 제거`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/ff0542f1330ef90834224f02acd68af0fd6c9421)

Gyuil-Hwnag replied:
`interceptor` 에 대한 내용 정리 아주 좋네요👍
`interceptor`는 그외에도 여러가지로 사용을 하는 경우가 있어서, 한번 내용에 대해서 알아두고 간다면 추후 개발을 할때 도움이 될 것 같습니다~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/repository/RecentlyViewedProductRepositoryImpl.kt:
```diff
@@ -30,9 +30,8 @@ class RecentlyViewedProductRepositoryImpl(
+                callback(product)
             }
         }
     }
```
`RecentlyViewedProductRepositoryImpl`에 대한 코멘트 입니다!
생성자를 외부에서도 접근이 가능하도록 열어두신 이유가 있었을까요~?

ijh1298 replied:
실수했습니다..😅 수정했습니다!

[`refactor: RecentlyViewedProductRepositoryImpl 생성자 매개변수 private으로 변경`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/d01f2c207e1f5fb1b80cb3b39c50714f24c43ba8)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/product/catalog/CatalogItem.kt:
```diff
@@ -1,9 +1,15 @@
+
+    data class LoadingStateProductItem(
+        val loadingState: LoadingState
+    ) : CatalogItem()
```
Loading에 대한 타입 추가 좋네요👍
다만, loadingState는 위에 남겨둔대로 존재에 대한 의문이 드는군요🥲

ijh1298 replied:
`LoadingStateProductItem`은 shimmer 라이브러리를 알기 전에 직접 아직 로딩이 되지 않은 아이템을 리사이클러뷰 아이템 뷰타입으로 지정해야겠다고 생각해서 만들었던 시도였는데.. 미처 삭제하지 못했었네요.😅

`LoadingStateProductItem`는 제거했습니다!

[`refactor: CatalogItem.LoadingStateProductItem 제거`](https://github.com/woowacourse/android-shopping-order/pull/105/commits/504c549f04e7f699f6d5e7e57b456e1b0ad56d02)

Gyuil-Hwnag reviewed (changes requested):
크림 안녕하세요~
이번 미션을 함께하게 된 두루라고 합니다. 잘부탁드립니다!

이번 미션 많이 어려우셨을텐데 정말 고생 많으셨습니다 👏
고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
화이팅입니다💪

Gyuil-Hwnag commented:
크림~ 아래의 플로우로 한다면, 봐야할 커밋만을 필터를 걸 수도 있습니다!
`PR > 좌측 commit > shift + click > 봐야할 커밋 드래그`
[수정 PR 필터](https://github.com/woowacourse/android-shopping-order/pull/105/files/cd91a8d44852defa75db6ecfb1e6ab339c6f00ef..0f46e955a6c9c9bf2997b832a793c8f48e79aa2d)

ijh1298 committed:
Merge branch 'step2' of https://github.com/ijh1298/android-shopping-order into step2

ijh1298 committed:
refactor: viewModel 선언부 by viewModels로 변경

ijh1298 committed:
refactor: CartAdapter quantityControlListener 매개변수 명칭 변경

ijh1298 committed:
refactor: CartViewModel의 버튼 이벤트 상수를 ButtonEvent enum으로 분리

ijh1298 committed:
refactor: 불필요한 with 제거

ijh1298 committed:
refactor: RecentlyViewedProductRepositoryImpl 생성자 매개변수 private으로 변경

ijh1298 committed:
refactor: Listener 인터페이스에 buttonEvent enum 적용

ijh1298 committed:
refactor: DetailActivity 함수 분리

ijh1298 committed:
refactor: 상품 수량 변경 및 화면 로직 개선

- DetailViewModel 초기화 시 상품 수량과 가격을 바로 설정하도록 변경
- 사용하지 않는 함수(setQuantity, setPriceSum, addToRecentlyViewedProduct) 제거
- DetailActivity에서 불필요한 LiveData observe 제거
- CatalogActivity에서 QuantityControlListener 이벤트 처리 방식 변경

ijh1298 committed:
refactor: 이미지 바인딩 로직 수정 및 코드 간소화

- imageUrl 바인딩 시 ProductUiModel 대신 String을 받도록 수정
- CatalogActivity 코드 간소화

ijh1298 committed:
refactor: CatalogViewModel page 변수 LiveData 제거

ijh1298 committed:
refactor: 수정한 ButtonEvent 상수 xml에 적용

ijh1298 committed:
refactor: 장바구니 수량 조절 interface를 model에서 id 기준으로 변경

ijh1298 committed:
refactor: deleteCartProduct 콜백으로 수정 및 적용

ijh1298 committed:
refactor: CartViewModel 페이지 번호 LiveData에서 일반 변수로 변경

ijh1298 committed:
refactor: LIMIT, OFFSET 사용한 쿼리문 의도대로 수정

ijh1298 committed:
refactor: CatalogItems 데이터 관리 방식 수정 및 UI 업데이트 로직 개선

- CatalogViewModel에서 `_catalogItems`를 `catalogItems`로 변경하고, `MutableList`로 관리하도록 수정했습니다.
- `_loadedCatalogItems`를 새로 추가하여 화면에 표시될 상품 목록을 관리하도록 변경했습니다.
- 상품 수량 감소 시, 장바구니에서 상품 삭제 후 UI 업데이트 로직을 수정했습니다.
- 페이지별 상품 로드 및 현재 페이지까지 상품 로드 로직에서 `catalogItems`를 직접 조작하고, `_loadedCatalogItems`를 업데이트하도록 변경했습니다.
- CatalogActivity에서 `onResume` 시 어댑터의 아이템을 초기화하고, `viewModel.loadedCatalogItems`를 관찰하여 상품 목록을 업데이트하도록 변경했습니다.
- ProductAdapter에서 `setItems` 메서드를 `addLoadedItems`로 변경하고, 아이템 추가 및 제거 시 `notifyItemRangeInserted`, `notifyItemRemoved`를 사용하도록 수정했습니다.
- 테스트 코드에서 `catalogItems` 대신 `loadedCatalogItems`를 참조하도록 수정했습니다.

ijh1298 committed:
refactor: 페이지네이션 버튼 업데이트 로직 개선

ijh1298 committed:
refactor: CatalogViewModel의 catalogItems 삭제 및 로딩 로직 변경

ijh1298 committed:
feat: 상품 이미지 로드 실패 시 대체 이미지 표시 및 크기 조정

ijh1298 committed:
fix: 장바구니 상품 수량 변경 및 삭제 시 cartItemId 사용하도록 수정

ijh1298 committed:
refactor: CartProductRepository interface 인자명 수정

ijh1298 committed:
refactor: CartViewModel 최근 본 상품 불러오는 함수 이름 변경

ijh1298 committed:
refactor: CartRecommendationFragment 내에서 상품 추가 가능하도록 수정

ijh1298 committed:
refactor: 페이지네이션 로직 제거 및 선택된 상품 개수 및 가격 갱신 로직 수정

ijh1298 committed:
refactor: Cart 초기 로딩이 아닐 시 Shimmer 적용 구분

ijh1298 committed:
refactor: CartViewModel LiveData 변수명 수정

ijh1298 committed:
refactor: Fragment 선택 방식을 최초 한번 결정하도록 수정 및 CartRecommendation 수량 조절 가능하도록 수정

- CartRecommendationFragment에서 수량 조절 시 가격 반영 되도록 수정

ijh1298 committed:
refactor: uid, id를 productId로 이름 변경

ijh1298 committed:
refactor: 최근 본 상품 목록 어댑터 ListAdapter로 변경 및 getProduct API 실패 처리 추가

- 최근 본 상품 목록 리사이클러뷰 작동하도록 수정

ijh1298 committed:
refactor: CartActionListener 인터페이스로 삭제 및 체크 이벤트 통합

ijh1298 committed:
feat: 전체 선택 체크박스 기능 구현

ijh1298 committed:
refactor: CartActivity에서 totalCount에 따라 Fragment를 교체하는 로직 개선

- 함수 분리 및 depth 개선
- Fragment 인스턴스 대신 클래스 전달

ijh1298 committed:
refactor: CartItem 삭제 시 ProductUiModel 대신 cartItemId 사용

ijh1298 committed:
refactor: ProductActionListener의 onLoadButtonClick 기본 구현 추가

ijh1298 committed:
refactor: Retrofit Interceptor에서 Accept 헤더 추가 및 서비스 인터페이스에서 헤더 제거

ijh1298 committed:
refactor: Fragment onDestroyView에서 binding 해제

ijh1298 committed:
refactor: ID, PW를 BuildConfig에서 가져오도록 수정 및 SharedPreference 제거

ijh1298 committed:
feat: RepositoryProvider, DependencyProvider 추가 및 ViewModelFactory 적용

- 각 ViewModelFactory에서 Repository를 직접 생성하는 대신 RepositoryProvider를 통해 주입받도록 변경
- Application 클래스에서 관리하던 Database 인스턴스를 DependencyProvider로 이동하여 관리
- ViewModelFactory 생성자에서 Application 의존성 제거

ijh1298 committed:
refactor: CatalogItem.LoadingStateProductItem 제거

ijh1298 reviewed (commented):
두루, 좋은 피드백 남겨주셔서 감사합니다!👍

첫 리뷰 제출 때도 버그나 구현이 부족한 부분이 많아 정상 작동이 되도록 우선적으로 고치고자 했습니다. 그 다음 리뷰 반영을 하다보니 못 고친 부분도 여전히 많았는데요..😭(Boolean 상태 값을 Activity에서 관리하지 말고 ViewModel에 MediatorLiveData 적용해보기, FragmentManager commit 하는 메서드 중복 줄이기, 테스트 코드 작성하지 못함.. 등)

전부 다 반영하고 리뷰 요청 드리기엔 늦을 거 같아서, 이후 리팩터링 때 계속해서 진행하겠습니다!

Gyuil-Hwnag commented on app/build.gradle.kts:
```diff
@@ -20,6 +29,17 @@ android {
         testInstrumentationRunnerArguments["runnerBuilder"] =
             "de.mannodermaus.junit5.AndroidJUnit5Builder"
+
+        buildConfigField(
+            "String",
+            "USER_ID",
+            "\"${localProperties["user.id"] ?: ""}\"",
+        )
+        buildConfigField(
+            "String",
+            "USER_PASSWORD",
+            "\"${localProperties["user.password"] ?: ""}\"",
+        )
```
locaProperties + buildConfig에 대한 세팅 정말 좋네요👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/ShoppingApplication.kt:
```diff
@@ -1,18 +1,11 @@
-
-        thread { DevMockServer.start() }
-        database = ShoppingDatabase.getInstance(applicationContext)
+        DependencyProvider.init(applicationContext)
     }
```
DataBase에 대한 초기화 내용 좋습니다👍
다만, 네이밍은 한번 고민이 들기는 했어요~
ShoppingDataBase 라는 네이밍이 드러나게 지어보는 것도 좋을 것 같다는 생각이 들었습니다!
다만, 이 부분이 DataBase 뿐만 아니라, 다른 인스턴스에 대해서도 초기화를 하는 로직이 들어가게 된다면 현재의 이름도 좋겠군요

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/ButtonEvent.kt:
```diff
@@ -0,0 +1,6 @@
+package woowacourse.shopping.cart
+
+enum class ButtonEvent {
+    INCREASE,
+    DECREASE,
+}
```
이 부분은 `INCREASE, DECREASE 이렇게 두개의 이벤트로 나누기 vs interface 에서 함수로 분리하기` 두가지중에서 개인적으로 제가 구현을 하는 것이었으면, 함수 두개로 분리하기를 택했을 수도 있을 것 같은데, 두개의 enum 으로 분리를 하셨던 이유가 있을까요?
두개를 비교해서, 이런점에서 enum 으로 분리를 하는 것이 좋았다!의 의견을 자유롭게 남겨주세요~💪
(다음 요청에서 코멘트로 남겨주셔도 좋을 것 같습니다!)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartActivity.kt:
```diff
@@ -3,86 +3,86 @@ package woowacourse.shopping.cart
-        val cartItemsWithPaginationBtn =
-            if (cartItems.isEmpty()) cartItems else cartItems + paginationItem
-        (binding.recyclerViewCart.adapter as CartAdapter).setCartItems(cartItemsWithPaginationBtn)
+        supportFragmentManager.commit {
+            setReorderingAllowed(true)
+            replace(
+                R.id.fragment_container_cart_selection,
+                CartRecommendationFragment::class.java,
+                null,
+            )
+        }
```
이 부분은 함수로 묶어분다면, 추후 다른 Fragment 들이 나올 때도 유용하게 사용을 해볼 수 있겠다는 생각도 들었습니다~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartAdapter.kt:
```diff
@@ -1,63 +1,48 @@
-        notifyItemChanged(index)
-    }
-
+    private val cartActionListener: CartActionListener,
+) : ListAdapter<ProductItem, RecyclerView.ViewHolder>(
```
ListAdapter + DiffUtils의 적용 아주 좋습니다~
이전 RecyclerView.Adapter와 비교했을 때 어떤 차이점이 있었을까요?!
크림의 자유로운 의견을 남겨주세요ㅎㅎ

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartRecommendationFragment.kt:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    private fun observeCartViewModel() {
+        val recommendProductAdapter =
+            binding.RecyclerViewCartRecommendation.adapter as ProductAdapter
```
이 부분에 대해서는 Adapter 자체를 전역변수로 생성을 한다면, 매번 Adapter 를 타입캐스팅 하는 부분에 대해서는 사라질 수도 있겠다는 생각도 들었습니다~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartRecommendationFragment.kt:
```diff
@@ -0,0 +1,88 @@
+        }
+    }
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+        _binding = null
+    }
```
onDestroyView() 에서 인스턴스에 대한 해제 정말 좋습니다👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/cart/CartSelectionFragment.kt:
```diff
@@ -0,0 +1,75 @@
+            )
+    }
+
+    private fun observeCartViewModel() {
+        val cartAdapter = binding.recyclerViewCart.adapter as CartAdapter
+        viewModel.cartProducts.observe(viewLifecycleOwner) { products ->
+            cartAdapter.submitList(products.toList())
+            viewModel.loadSelectAllState()
+        }
+        viewModel.loadingState.observe(viewLifecycleOwner) { changeShimmerState(it) }
+    }
```
이 부분은 코드레빗이 좋은 내용을 남겨준 것 같군요~
[코드레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/105/files/cd91a8d44852defa75db6ecfb1e6ab339c6f00ef..504c549f04e7f699f6d5e7e57b456e1b0ad56d02#r2122952314)

방법들은 MediatorLiveData 를 사용하거나, ViewModel 에서 호출하는등 여러가지 방법이 있을 것 같군요~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/repository/RecentlyViewedProductRepository.kt:
```diff
@@ -4,7 +4,7 @@ import woowacourse.shopping.data.entity.CartProductEntity
 
 interface RecentlyViewedProductRepository {
-    fun insertRecentlyViewedProductUid(uid: Int)
+    fun insertRecentlyViewedProductId(uid: Int)
```
이 부분은 변수명 수정을 깜빡하셨나 보군요~😅

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/repository/RemoteCartProductRepositoryImpl.kt:
```diff
@@ -0,0 +1,245 @@
+import woowacourse.shopping.product.catalog.ProductUiModel
+
+class RemoteCartProductRepositoryImpl : CartProductRepository {
+    val retrofitService = RetrofitProductService.INSTANCE.create(CartItemService::class.java)
+
```
엇! retrofitService 부분이 외부로 드러나게 하는걸 막는걸 깜빡하셨나 보군요~

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/repository/RemoteCartProductRepositoryImpl.kt:
```diff
@@ -0,0 +1,245 @@
+                        }
+                    }
+
+                    override fun onFailure(
+                        call: Call<Void>,
+                        t: Throwable,
+                    ) {
+                        println("error : $t")
```
이 부분은 코드레빗의 코멘트를 한번 살펴보는 것도 좋을 것 같아서 링크 첨부해두었습니다~
[코드레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/105/files/cd91a8d44852defa75db6ecfb1e6ab339c6f00ef..504c549f04e7f699f6d5e7e57b456e1b0ad56d02#r2122952396)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/shopping/data/service/RetrofitProductService.kt:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    private val authInterceptor =
+        Interceptor { chain ->
+            val original: Request = chain.request()
+            val requestWithAuth: Request =
+                original
+                    .newBuilder()
+                    .header("Authorization", "Basic $base64Credentials")
+                    .header("accept", "*/*")
+                    .build()
+            chain.proceed(requestWithAuth)
+        }
```
interceptor 의 활용 너무 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/res/layout/activity_cart.xml:
```diff
@@ -5,23 +5,93 @@
+            android:layout_height="wrap_content"
+            tools:text="184,000원"
+            android:text="@{@string/format_price(vm.totalPrice)}"
+            style="@style/CartPageOrderStyle"
+            android:layout_marginEnd="12dp"
```
styles.xml 의 활용 너무 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/res/layout/cart_item.xml:
```diff
@@ -26,17 +26,31 @@
+            android:id="@+id/checkbox_selection"
+            android:minWidth="48dp"
+            android:minHeight="48dp"
+            android:onClick="@{() -> cartActionListener.onCheckToggle(cartProduct)}"
+            android:checked="true"
```
이 부분은 코드레빗의 코멘트도 한번 살펴보면 좋을 것 같다는 생각이 들었습니다~
[코드레빗 코멘트](https://github.com/woowacourse/android-shopping-order/pull/105/files/cd91a8d44852defa75db6ecfb1e6ab339c6f00ef..504c549f04e7f699f6d5e7e57b456e1b0ad56d02#r2122952388)

~~~kotlin
             android:minHeight="48dp"
             android:onClick="@{() -> cartActionListener.onCheckToggle(cartProduct)}"
-            android:checked="true"
+            android:checked="@{cartProduct.isSelected}"
             android:layout_width="wrap_content"
~~~
이 부분은 BindingAdapter를 연습해볼 수 있는 기회일 것 같군요!

Gyuil-Hwnag reviewed (approved):
크림 안녕하세요~
해당 미션 너무 고생 많으셨습니다!
리뷰 반영 정말 잘 해주셨네요💯
해당 미션은 다음 미션 진행을 위하여 머지하도록 하겠습니다.

이번 미션 Server 와의 연동등 많이 어려우셨을텐데, 너무 잘해주셨습니다👍
도움이 될만한 코멘트 달았으니 확인하여 다음 미션에서 반영해주셔도 좋을 것 같습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
다음 미션도 화이팅입니다💪
